src/crafting/CraftingManager.php

```diff
@@ -197,6 +204,23 @@ public function getPotionContainerChangeRecipes() : array{
 		return $this->potionContainerChangeRecipes;
 	}
 
+	/**
+	 * @return SmithingRecipe[]
+	 * @phpstan-return list<SmithingRecipe>
+	 */
+	public function getSmithingRecipes() : array{
+		return $this->smithingRecipes;
+	}
+
+	public function getSmithingRecipeFromIndex(int $index) : ?SmithingRecipe{
+		$craftingOffset = count($this->craftingRecipeIndex) - 1;
```

Comment:
This doesn't look reliable. If new crafting recipes are registered, the offset will change.

Reply:
Smithing recipes are still behind crafting recipes, the relative position of the smithing recipes will not change. Also all caches are regenerated once a new recipe is registered, so I dont get your point.

Reply:
yeah it seems possible, what should I do though? 🤷🏼

Reply:
> If a recipe is registered after a player joins, the recipe IDs they send back to the server won't match up with this.

Which plugins add recipes after the join? but anyway, how would you solve that? @dktapps

Reply:
Not sure tbh. It seems stupid to me that these recipes share the same ID space with crafting recipes, since they can't be crafted in a crafting table. However, I expect that this is required by the client.

Perhaps a bitflag on the recipe ID would work, or a large fixed offset.

Reply:
We had a similar problem with enchanting options. An offset of 100,000 was used to prevent conflicts there. To use the same approach here you'd need to go with 200k or something like that.

We'll need a better solution for this in the longer term, but that's a problem for another PR.